File: android-offline/src/com/apolos/shieldlite/PacketUtils.java
package com.apolos.shieldlite;

/** Minimal IPv4/UDP/DNS byte-manipulation helpers for the DNS-filtering tunnel. */
final class PacketUtils {

    static final int PROTO_UDP = 17;

    private PacketUtils() { }

    static int ipVersion(byte[] pkt) { return (pkt[0] & 0xF0) >>> 4; }

    static int ihlBytes(byte[] pkt) { return (pkt[0] & 0x0F) * 4; }

    static int protocol(byte[] pkt) { return pkt[9] & 0xFF; }

    static int udpDstPort(byte[] pkt, int ihl) {
        return ((pkt[ihl + 2] & 0xFF) << 8) | (pkt[ihl + 3] & 0xFF);
    }

    static int udpSrcPort(byte[] pkt, int ihl) {
        return ((pkt[ihl] & 0xFF) << 8) | (pkt[ihl + 1] & 0xFF);
    }

    static byte[] udpPayload(byte[] pkt, int ihl, int length) {
        int start = ihl + 8;
        byte[] out = new byte[Math.max(0, length - start)];
        System.arraycopy(pkt, start, out, 0, out.length);
        return out;
    }

    static String dnsQueryName(byte[] dns) {
        if (dns.length < 13) return null;
        int pos = 12;
        StringBuilder sb = new StringBuilder();
        while (pos < dns.length) {
            int len = dns[pos] & 0xFF;
            if (len == 0) break;
            if ((len & 0xC0) != 0) return null;
            pos++;
            if (pos + len > dns.length) return null;
            if (sb.length() > 0) sb.append('.');
            for (int i = 0; i < len; i++) sb.append((char) (dns[pos + i] & 0xFF));
            pos += len;
        }
        return sb.toString().toLowerCase();
    }

    static byte[] buildNxDomain(byte[] query) {
        int pos = 12;
        while (pos < query.length) {
            int len = query[pos] & 0xFF;
            if (len == 0) { pos++; break; }
            if ((len & 0xC0) != 0) { pos += 2; break; }
            pos += 1 + len;
        }
        int end = Math.min(pos + 4, query.length);
        byte[] out = new byte[end];
        System.arraycopy(query, 0, out, 0, end);
        out[2] = (byte) 0x81;
        out[3] = (byte) 0x83;
        out[6] = 0; out[7] = 0;
        out[8] = 0; out[9] = 0;
        out[10] = 0; out[11] = 0;
        return out;
    }

    static byte[] buildUdpResponse(byte[] request, int requestIhl, byte[] payload) {
        int ipHeaderLen = 20;
        int udpLen = 8 + payload.length;
        int total = ipHeaderLen + udpLen;
        byte[] out = new byte[total];

        out[0] = 0x45;
        out[1] = 0;
        putShort(out, 2, total);
        putShort(out, 4, 0);
        putShort(out, 6, 0x4000);
        out[8] = 64;
        out[9] = (byte) PROTO_UDP;
        System.arraycopy(request, 16, out, 12, 4);
        System.arraycopy(request, 12, out, 16, 4);
        putShort(out, 10, 0);
        putShort(out, 10, checksum(out, 0, ipHeaderLen));

        int srcPort = udpDstPort(request, requestIhl);
        int dstPort = udpSrcPort(request, requestIhl);
        putShort(out, 20, srcPort);
        putShort(out, 22, dstPort);
        putShort(out, 24, udpLen);
        putShort(out, 26, 0);
        System.arraycopy(payload, 0, out, 28, payload.length);
        return out;
    }

    private static void putShort(byte[] buf, int off, int value) {
        buf[off] = (byte) ((value >>> 8) & 0xFF);
        buf[off + 1] = (byte) (value & 0xFF);
    }

    private static int checksum(byte[] buf, int off, int len) {
        int sum = 0;
        int i = off;
        int end = off + len;
        while (i + 1 < end) {
            sum += ((buf[i] & 0xFF) << 8) | (buf[i + 1] & 0xFF);
            i += 2;
        }
        if (i < end) sum += (buf[i] & 0xFF) << 8;
        while ((sum >>> 16) != 0) sum = (sum & 0xFFFF) + (sum >>> 16);
        return (~sum) & 0xFFFF;
    }
}

File: android/app/src/main/java/com/apolos/shield/vpn/PacketUtils.kt
package com.apolos.shield.vpn

import java.nio.ByteBuffer

/**
 * Minimal IPv4 / UDP helpers for the DNS-filtering tunnel. Only what the DNS
 * path needs: parse a query name, and rebuild an IPv4+UDP response packet with
 * correct checksums so the OS resolver accepts it.
 */
object PacketUtils {

    const val PROTO_UDP = 17

    fun ipVersion(pkt: ByteArray): Int = (pkt[0].toInt() and 0xF0) ushr 4
    fun ihlBytes(pkt: ByteArray): Int = (pkt[0].toInt() and 0x0F) * 4
    fun protocol(pkt: ByteArray): Int = pkt[9].toInt() and 0xFF

    fun udpDstPort(pkt: ByteArray, ihl: Int): Int =
        ((pkt[ihl + 2].toInt() and 0xFF) shl 8) or (pkt[ihl + 3].toInt() and 0xFF)

    fun udpSrcPort(pkt: ByteArray, ihl: Int): Int =
        ((pkt[ihl].toInt() and 0xFF) shl 8) or (pkt[ihl + 1].toInt() and 0xFF)

    /** UDP payload (the DNS message) as a fresh array. */
    fun udpPayload(pkt: ByteArray, ihl: Int, length: Int): ByteArray {
        val start = ihl + 8
        return pkt.copyOfRange(start, length)
    }

    /** Reads the queried host name from a DNS query message. */
    fun dnsQueryName(dns: ByteArray): String? {
        if (dns.size < 13) return null
        var pos = 12
        val sb = StringBuilder()
        while (pos < dns.size) {
            val len = dns[pos].toInt() and 0xFF
            if (len == 0) break
            if (len and 0xC0 != 0) return null // compression not expected in a query
            pos++
            if (pos + len > dns.size) return null
            if (sb.isNotEmpty()) sb.append('.')
            for (i in 0 until len) sb.append((dns[pos + i].toInt() and 0xFF).toChar())
            pos += len
        }
        return sb.toString().lowercase()
    }

    /** Builds an NXDOMAIN reply for a blocked query (works for any record type). */
    fun buildNxDomain(query: ByteArray): ByteArray {
        // Truncate to header (12) + the single question, dropping any EDNS OPT so
        // the response is well-formed. Question = name labels + qtype(2) + qclass(2).
        var pos = 12
        while (pos < query.size) {
            val len = query[pos].toInt() and 0xFF
            if (len == 0) { pos++; break }
            if (len and 0xC0 != 0) { pos += 2; break } // pointer (unexpected in query)
            pos += 1 + len
        }
        val end = (pos + 4).coerceAtMost(query.size) // + qtype + qclass
        val out = query.copyOf(end)
        out[2] = 0x81.toByte()   // QR=1, Opcode=0, RD=1
        out[3] = 0x83.toByte()   // RA=1, rcode=3 (NXDOMAIN)
        out[6] = 0; out[7] = 0   // ANCOUNT
        out[8] = 0; out[9] = 0   // NSCOUNT
        out[10] = 0; out[11] = 0 // ARCOUNT
        return out
    }

    /**
     * Wraps [payload] into an IPv4+UDP packet that is the reply to [request]:
     * addresses and ports swapped, lengths and checksums recomputed.
     */
    fun buildUdpResponse(request: ByteArray, requestIhl: Int, payload: ByteArray): ByteArray {
        val ipHeaderLen = 20
        val udpLen = 8 + payload.size
        val total = ipHeaderLen + udpLen
        val out = ByteArray(total)

        // ---- IPv4 header ----
        out[0] = 0x45                      // version 4, IHL 5
        out[1] = 0                         // DSCP/ECN
        putShort(out, 2, total)            // total length
        putShort(out, 4, 0)                // identification
        putShort(out, 6, 0x4000)          // flags = don't fragment
        out[8] = 64                        // TTL
        out[9] = PROTO_UDP.toByte()
        // swap src/dst IPs
        System.arraycopy(request, 16, out, 12, 4) // dst -> src
        System.arraycopy(request, 12, out, 16, 4) // src -> dst
        putShort(out, 10, 0)               // checksum placeholder
        putShort(out, 10, checksum(out, 0, ipHeaderLen))

        // ---- UDP header ----
        val srcPort = udpDstPort(request, requestIhl) // reply from the port we were asked on (53)
        val dstPort = udpSrcPort(request, requestIhl)
        putShort(out, 20, srcPort)
        putShort(out, 22, dstPort)
        putShort(out, 24, udpLen)
        putShort(out, 26, 0)               // checksum optional for IPv4; 0 = disabled
        System.arraycopy(payload, 0, out, 28, payload.size)
        return out
    }

    private fun putShort(buf: ByteArray, off: Int, value: Int) {
        buf[off] = ((value ushr 8) and 0xFF).toByte()
        buf[off + 1] = (value and 0xFF).toByte()
    }

    private fun checksum(buf: ByteArray, off: Int, len: Int): Int {
        var sum = 0
        var i = off
        val end = off + len
        while (i + 1 < end) {
            sum += ((buf[i].toInt() and 0xFF) shl 8) or (buf[i + 1].toInt() and 0xFF)
            i += 2
        }
        if (i < end) sum += (buf[i].toInt() and 0xFF) shl 8
        while (sum shr 16 != 0) sum = (sum and 0xFFFF) + (sum ushr 16)
        return sum.inv() and 0xFFFF
    }
}
